### PR TITLE
Truncate SES message data to CloudWatch

### PR DESCRIPTION
**Pull request by ivankristianto**
Fixes: #371 

This PR is similar with the PR I made in https://github.com/humanmade/aws-xray/pull/77
With solve the same issue where the data send to cloudwatch could be more than 64Kb and thus we cannot see the email log if the email has attachment.

For logging purpose we don't need to store email attachment in the cloudwatch, it's enough information to see if the email header and delivery status.

So this PR truncate the message down.
Let me know what you think? @rmccue @roborourke @joehoyle

### CODE REVIEW

File: inc/ses_to_cloudwatch/namespace.php [hunk 1]
```diff
@@ -43,3 +53,21 @@ function on_error_sending_message( Exception $error, $message ) {
...
+ * Truncate string to given maximum size.
+ *
+ * @param string $message     String message to truncate.
+ * @param int    $max_size    Maximum size in bytes, default 5KB.
+ * @param string $replacement String replacement.
+ *
+ * @return string
+ */
+function truncate_message( string $message, int $max_size = 5 * 1024, string $replacement = '…' ) : string {
+	if ( mb_strlen( $message ) < $max_size ) {
```
Comment (joehoyle) [inline inc/ses_to_cloudwatch/namespace.php:67]: If you are counting length to be a proxy for size, this should be `strlen` not `mb_strlen`. A string that is `mb_strlen` of 1 can be a size of 3bytes for example.

**Review by joehoyle — CHANGES_REQUESTED** [review]

File: inc/ses_to_cloudwatch/namespace.php [hunk 2]
```diff
@@ -43,3 +53,21 @@ function on_error_sending_message( Exception $error, $message ) {
...
+/**
+ * Truncate string to given maximum size.
+ *
+ * @param string $message     String message to truncate.
+ * @param int    $max_size    Maximum size in bytes, default 5KB.
+ * @param string $replacement String replacement.
+ *
+ * @return string
+ */
+function truncate_message( string $message, int $max_size = 5 * 1024, string $replacement = '…' ) : string {
```
Comment (joehoyle) [inline inc/ses_to_cloudwatch/namespace.php:66]: This seems maybe `truncate_string` is more appropriate, as from the perspective of this function, it doesn't need to be a "message"

**Review by joehoyle — APPROVED** [review]

**Comment by joehoyle** [comment]
Handing off to @roborourke for final review

File: inc/ses_to_cloudwatch/namespace.php [hunk 3]
```diff
@@ -43,3 +53,21 @@ function on_error_sending_message( Exception $error, $message ) {
 		'message' => $message,
 	] ) );
 }
+
+/**
+ * Truncate string to given maximum size.
+ *
+ * @param string $string      String message to truncate.
+ * @param int    $max_size    Maximum size in bytes, default 5KB.
+ * @param string $replacement String replacement.
```
Comment (roborourke) [inline inc/ses_to_cloudwatch/namespace.php:62]: We don't align docblock params any more, just prevents unnecessary code changes when adding / changing a parameter in future.
```suggestion
 * @param string $string String to truncate.
 * @param int $max_size Maximum size in bytes, default 5KB.
 * @param string $replacement String replacement.
```

File: inc/ses_to_cloudwatch/namespace.php [hunk 4]
```diff
@@ -43,3 +53,21 @@ function on_error_sending_message( Exception $error, $message ) {
...
+ *
+ * @return string
+ */
+function truncate_string( string $string, int $max_size = 5 * 1024, string $replacement = '…' ) : string {
+	if ( strlen( $string ) < $max_size ) {
+		return $string;
+	}
+
+	// Truncate string in the middle.
+	return substr_replace( $string, $replacement, $max_size / 2, strlen( $string ) - $max_size + strlen( $replacement ) );
```
Comment (roborourke) [inline inc/ses_to_cloudwatch/namespace.php:72]: Took me a while to grok what this is doing! Nice work

**Review by roborourke — APPROVED** [review]

**Pull request merged**